### PR TITLE
fix(export): one predicate for "will this line be in the file" — a plain full export could ship a relationship naming an entity it never wrote

### PR DESCRIPTION
**Pull request by BIMvoice**
> **Draft.** An adversarial review of this branch found two HIGH defects and several false claims in this description. They are fixed in `2d4b5ba` and the description below is corrected; the findings, with reproductions, are in a comment on this thread. One design question — the walk/output predicate divergence — is open and is the maintainer's call, which is why this is back in draft.

## The bug

A plain full STEP export — **no `visibleOnly`, no deletions, no overlay** — emits a relationship naming an entity it never writes. Measured on a store whose record for `#2` claims a byte range the source buffer cannot serve:

```
#1=IFCWALL('0walA0000000000000000',$,'KeptWall',$,$,$,$,$);
#3=IFCRELCONTAINEDINSPATIALSTRUCTURE('0cont0000000000000000',$,$,$,(#1,#2),#4);
#4=IFCBUILDINGSTOREY('0stor0000000000000000',$,'S',$,$,$,$,$,$,0.);
```

`#3` names `#2`; there is no `#2=` line. Strict viewers reject such a file; lenient ones fall the geometry back to the origin — this package's own stated position in `entity-iteration.ts`.

Reproduction is `unreadable-ref-dangling.test.ts`, shown RED before the fix (`expected [ 2 ] to deeply equal []`).

## Scope, stated up front

**This targets a corrupt-input edge case, not everyday exports.** On a well-formed model the output does not change. Measured on `tests/models/AB22.ifc` — plain, under `visibleOnly`, and under `includeGeometry: false` — the bytes are identical to `main` but for the header timestamp. Anyone reading "one predicate for will this line be in the file" should know that up front.

## Root cause: two predicates for one question

`willBeEmitted` (`step-exporter.ts`) recognises **seven** reasons an entity's line never lands in the file:

| # | Reason | Old filter predicate | Now |
|---|---|---|---|
| 1 | Outside the `visibleOnly` closure | only for hidden **products** | reachable by the predicate, **no test** |
| 2 | Hidden product | yes | yes |
| 3 | Tombstoned | yes | yes |
| 4 | Never existed | no | deliberately not — see below |
| 5 | **Unreadable source ref** (#2491) | **no — this is the live one** | **fixed, with tests** |
| 6 | Geometry excluded by `includeGeometry: false` | no | reachable by the predicate, **not meaningfully covered** |
| 7 | `deltaOnly` / `exportPropertiesOnly` skipping the pass | n/a | n/a — see below |

The relationship-reference filter consumed `isExcludedFromRelationshipRefs`, which answered `hiddenProductIds.has(id) || effective.isDeleted(id)`. On top of that sat a **second** gate, `mayNameExcludedRefs`, which ran the filter only when hidden products existed or an overlay was active — false for the export above, so the filter never ran at all.

## What changed

Both output-line filter sites now consume **`isOmittedFromOutput`**, derived from `willBeEmitted`, and `mayNameExcludedRefs` is deleted.

Reason by reason, without rounding up:

- **(5) is fixed**, and is the only reason with a reproduction and behavioural tests.
- **(1) is reachable by the predicate but has no test.** No fixture has produced it either: a relationship the closure walk refuses to bridge is one the output filter already withholds, because the walk predicate is a subset of the output predicate.
- **(6) is reachable by the predicate but is not meaningfully covered.** The filter only rewrites `IFCREL*` lines, and geometry is named from a product's `Representation` / `ObjectPlacement` slot. `includeGeometry: false` on `AB22.ifc` yields **80 dangling refs on this branch and on `main`**, byte-identical output. An earlier revision of this description claimed (1) and (6) were "covered too"; that was wrong and is retracted.
- **(7) is not applicable.** `deltaOnly` skips the source-iteration pass wholesale, and `willBeEmitted` answers `true` for a source entity there because a delta patches a file that already holds the line. Note this is also a place where `isOmittedFromOutput` and actual emission disagree, and the predicate's docstring now says so.
- **(4) is deliberately excluded**, and is the one place the filter is narrower than `willBeEmitted`. A `#999` already dangling in the **input** file is not this export's doing, and `filterHiddenRefsFromRelationshipLine` withholds a whole relationship when an excluded id sits in a bare scalar — so treating it as an exclusion **deleted a visible element's pset over somebody else's corrupt file**. That is the harm #2637 was about. Hence `isOmittedFromOutput` reads `(effective.has(id) || effective.isDeleted(id)) && !willBeEmitted(id)`.

## The default export path can now drop a relationship — and now says so

Withholding beats shipping a dangling `#N`, but it is a real edit: when the omitted id sits in a single-valued attribute, or is a set's only member, the **whole relationship line is dropped**, and anything else that relationship named loses the association. This reaches a plain full export with no options set at all.

That was silent. Every such drop now pushes a message onto `stats.warnings` naming the relationship and its type, pinned by a test that is RED with the push removed, plus a companion test asserting a *rewritten* relationship stays silent.

## The closure walk stays separate — and that is a real departure

Using `willBeEmitted` at the closure walk is circular: it reads `allowedEntityIds`, which is what `collectReferencedEntityIds` produces. Wired in and measured, it does not even evaluate:

```
ReferenceError: Cannot access 'willBeEmitted' before initialization
 ❯ src/step-exporter.ts:436:25
 ❯ relationshipRefsSurviveExclusion src/reference-collector.ts:673:16
 ❯ collectReferencedEntityIds src/reference-collector.ts:482:12
```

So the walk keeps the narrower `isRefExcludedDuringClosureWalk`. **This breaks the contract #2637 was closed on** — `reference-collector.ts` documented the bridge as taking the caller's own OUTPUT predicate, "not two expressions that happened to agree". The doc, and the stale `isExcludedFromRelationshipRefs` references in `merged-exporter.ts` and `step-exporter.test.ts`, are corrected rather than left to read as still true.

The divergence is **ordered, not symmetric**: everything the walk excludes, the output excludes too, so the #2548 closure leak cannot return. The reverse gap is open and observable — for an unreadable ref the walk bridges through a relationship the output then withholds, leaving an orphan pset. A test now pins exactly that, so it is recorded rather than found later. **Whether this is the right shape is the open question on this PR**, and the reason it is a draft.

## How the wiring is pinned

By behaviour only. Each test feeds an input on which the walk predicate and the output predicate give different answers and checks the emitted bytes. Substituting the walk predicate at either call site turns five of six tests RED; gutting `willBeEmitted` to `return true` turns three RED.

An earlier revision carried a source-text "drift guard" that `readFileSync`'d `step-exporter.ts`. It is **deleted, not allowlisted**: it broke the repo's own `Check for new source-text assertions` gate (the pattern removed from #2393 today), and it was vacuous — with `willBeEmitted` gutted to `return true`, all four of its assertions stayed green.

## Second, smaller fix: a wrong justification in `source-ref-bounds.ts`

The doc justified exempting the incidental readers (`getPropertySetName` and siblings) with "a clamped, empty decode already yields no match, which is the same answer". **That reason is false**, and an earlier revision of this branch got the correction half wrong too: it said an overrunning END "is genuinely benign". It is not.

A clamp does not empty a range; it moves an endpoint onto a real file byte, and the window that survives still holds somebody else's record. Both directions produce a confidently wrong answer:

```
negative offset:  getPropertySetName(#2)  = 'SetA'      <- #1's name
overrunning end:  getPropertyIdsInSet(#1) = [201,202]   <- #2's members ( #1's are [101,102] )
                  isReadableSourceRef(#1) = false
```

The `$`-anchored patterns (`getPropertyIdsInSet`, `getRelatedPropertySet`) match at the end of the **clamped** window, i.e. against whatever record the file ends on. `getPropertySetName` is right there only by luck, because its pattern is unanchored.

The overrun is also the shape that is **reachable** — it is the same corrupt-store shape this PR is about, a ref claiming bytes the source cannot serve — where the negative offset is not: `OVERLAY_BYTE_OFFSET` (`-1`) is the only negative offset in the repo, all three sites that write it (`store-editor.ts` `addEntity`, `effective-index.ts` `get` and `[Symbol.iterator]`) pair it with `byteLength: 0`, and it is synthesised on read by the *effective* index, never written into `dataStore.entityIndex.byId`, which is the index those readers consult. Those facts are pinned by tests, and the pairing guard was mutation-tested (changing `byteLength: 0` to `20` in `store-editor.ts` turns it RED).

So the exemption is a live defect rather than a safe simplification, and it predates this branch. **Fixing it was #2678**, which gates those readers on `isReadableSourceRef` — the same predicate this branch's emission pass uses — with the probe and the tests. **#2678 has since merged**, and this branch's merge of `main` took the post-#2678 docstring, so the prose fix this section describes as pending is already in `main`: `source-ref-bounds.ts` carries **no diff** in this PR (only `source-ref-bounds.test.ts` does). The argument deliberately stays in `source-ref-bounds.ts` so that `step-exporter.ts`'s `entityLineText` docstring can cite it rather than carry a second copy that has to stay true in two places.

## Known cost

`filterHiddenRefsFromRelationshipLine` now runs on every `IFCREL*` line rather than behind the `mayNameExcludedRefs` gate. Measured on `ara3d/schependomlaan.ifc` (714,485 entities), export only, 5 timed runs after a warm-up: median **461 ms → 498 ms**, about **+8%**.

**That figure understates the cost.** A 21-rep interleaved randomized remeasurement of this same code puts it at **+13–15%** (523.4 / 522.6 ms against a 454.9 / 461.8 ms floor). The +8% is the number the original framing was built on, so it is corrected here rather than left standing. The maintainer has since ruled that cost not worth paying for a defect only reachable on corrupt input, and proposed a cheaper gate measured at **+4.1%** (`proposal/emission-gate-for-2668`) — not adopted into this branch, and the direction of this PR is open pending that.

## Verification

- `pnpm --filter @ifc-lite/export test` — 702 passed, 30 skipped; 1 file fails only on an unbuilt `@ifc-lite/wasm` in this environment (`lod1-generator.test.ts`), unrelated to this diff
- `pnpm --filter @ifc-lite/export typecheck` — OK (50 test files)
- `node scripts/check-source-text-assertions.mjs` — OK (6 allowlisted, 0 new)
- `node scripts/check-test-wiring.mjs` — OK
- `pnpm exec oxlint packages/export/src` — no new findings

Every added test was shown RED via a surgical mutation. No assertion weakened, no fixture adjusted, no skip added, no allowlist entry.

## References

- #2637 — the prior instance of this class in this file, and the contract this branch departs from.
- #2398 — where this was found.
- #2491 — introduced `isReadableSourceRef` (omission reason 5) and `source-ref-bounds.ts`.

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed STEP exports that contained unreadable or dangling source references.
  * Relationships now correctly omit, rewrite, or withhold invalid references without producing broken files.
  * Improved handling for hidden, deleted, geometry-excluded, and unreadable entities.
  * Added warnings when relationships cannot be represented in the exported file.
  * Corrected source-reference handling for negative offsets and overlay-created content.
* **Tests**
  * Expanded coverage for visibility, geometry exclusion, unreadable references, relationship filtering, and warning behavior.

### CODE REVIEW

File: .changeset/step-export-unreadable-ref-dangling.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+**Scope: this targets a corrupt-input edge case, not everyday exports.** On a well-formed model nothing changes. Measured on `tests/models/AB22.ifc`, plain and under `visibleOnly` and `includeGeometry: false`, the output is byte-identical to the previous release but for the header timestamp. The `includeGeometry: false` export still carries 80 dangling refs, exactly as before: the filter only rewrites `IFCREL*` lines, and geometry is named from products' `Representation` / `ObjectPlacement` slots, which it does not touch.
+
+The cause was two predicates for one question. `willBeEmitted` recognises seven reasons a line never lands in the file, while the relationship-reference filter consumed a separate predicate that answered for three of them (hidden product, tombstoned, never existed) — and a second gate in front of the filter suppressed it entirely unless hidden products or an overlay were present. Both relationship-emission passes now filter on the negation of `willBeEmitted`, and that gate is gone.
```
[the remark this answers is not in the document]
Comment (BIMvoice) [reply .changeset/step-export-unreadable-ref-dangling.md]: Fixed in d78c4e042 — and the imprecision was worse than "describe it precisely", so thank you for the pull.

Two defects in that one paragraph:

**1. "the negation of `willBeEmitted`" is not what the code does.** The filter consumes `isOmittedFromOutput`, which is `(effective.has(id) || effective.isDeleted(id)) && !willBeEmitted(id)` — that negation narrowed to ids the model actually has, or has tombstoned. The narrowing is load-bearing, not tidiness: `willBeEmitted` also answers NO for an id neither the file nor the session ever had, so under the *plain* negation a `#999` already dangling in the **input** file would count as omitted. `filterHiddenRefsFromRelationshipLine` withholds a whole relationship when an omitted id sits in a bare scalar, so that would drop the entire relationship — and every association it also carried — deleting a visible element's pset over somebody else's corrupt file. That is exactly the harm #2637 was about. The changeset as written described a predicate that would reintroduce it.

**2. "three of them (hidden product, tombstoned, never existed)" was a miscount.** The predicate it replaced was

```ts
const isExcludedFromRelationshipRefs = (id: number): boolean =>
  (hiddenProductIds !== null && hiddenProductIds.has(id)) || effective.isDeleted(id);
```

Two reasons — hidden product, tombstoned — and pointedly **not** "never existed", which is the one the new predicate still deliberately declines to treat as an omission. 41cdce830's commit message had this right ("answered for two of them"); the changeset prose had drifted from it.

The changeset now gives the expression verbatim, states the two reasons the old predicate covered, and spends a paragraph on why the scope qualifier exists — including that a pre-existing dangling ref stays out of scope and ships as it arrived, and that the tombstone arm is spelled out separately because `effective.has` is false for a tombstone while deleting an entity *is* this session's doing.

I also swept for the same drift in code prose and corrected three further sites that repeated one or other version of it: the `isOmittedFromOutput` JSDoc in `step-exporter.ts`, the file header of `unreadable-ref-dangling.test.ts`, and the `isRefExcluded` param doc in `reference-collector.ts`.

File: packages/export/src/source-ref-bounds.test.ts [hunk 2]
```diff
@@ -212,3 +213,68 @@ describe('the exporter never writes a reference to a line it could not emit', ()
...
+    // sites that write it — `store-editor.ts` `addEntity`, and
+    // `effective-index.ts` `get` / `[Symbol.iterator]` — pair it with
+    // `byteLength: 0`, which is the case the clamp really does answer as empty.
+    expect(OVERLAY_BYTE_OFFSET).toBeLessThan(0);
+    const store = await parseBase();
+    const created = new StoreEditor(store, new MutablePropertyView(null, 'test-model'))
+      .addEntity('IFCWALL', [null, null, null, null, null, null, null, null]);
+    expect(created.byteOffset).toBe(OVERLAY_BYTE_OFFSET);
+    expect(created.byteLength).toBe(0);
+  });
```
[the remark this answers is not in the document]
Comment (BIMvoice) [reply packages/export/src/source-ref-bounds.test.ts:295]: Fixed in d78c4e042. This was a real gap, and I confirmed it was more than theoretical before closing it.

**First, the reassuring part:** all three producers still pair the negative offset with a zero length. Verified by reading each site, not inferred:

- `packages/mutations/src/store-editor.ts:166` — `addEntity`'s returned ref: `byteOffset: OVERLAY_BYTE_OFFSET, byteLength: 0`
- `packages/export/src/effective-index.ts:282` — `OverlayIndex.get`: same pairing
- `packages/export/src/effective-index.ts:304` — `OverlayIndex[Symbol.iterator]`: same pairing

`OVERLAY_BYTE_OFFSET` is the only negative offset in the repo (`grep` over `packages/**/*.ts` excluding tests returns those three writers plus the declaration and two re-exports). So the wrong answer stays unreachable.

**Then the gap.** The test now asserts all three, and the iterator site matters separately because it builds its own ref object literal rather than delegating to `get` — `get` passing says nothing about it. There is also a closing loop asserting `byteOffset < 0 ⇒ byteLength === 0` over everything the iterator yields, which is the invariant in the form the incidental readers actually depend on.

**RED first**, one producer at a time, by changing a single `byteLength: 0` at the producer and restoring:

- `effective-index.ts:282` → `byteLength: 5` → `AssertionError: expected 5 to be +0`, 1 failed | 9 passed
- `effective-index.ts:304` → `byteLength: 7` → `AssertionError: expected 7 to be +0`, 1 failed | 9 passed

And the measurement that shows the gap was real rather than cosmetic: with the **iterator** site broken, the *previous* revision of this test file still reported `Tests 10 passed (10)`. The prose named three producers; only one could have caught this.

Both files restored; `packages/export` is 48 files / 705 tests green, typecheck clean, `check-source-text-assertions` OK (6 allowlisted, 0 new).

File: packages/export/src/reference-collector.ts [hunk 3]
```diff
@@ -612,10 +635,12 @@ export function filterHiddenRefsFromRelationshipLine(
   if (!match) return line;
   const [, prefix, argsText, suffix] = match;
   const attrs = splitTopLevelArgs(argsText);
+  const entityType = prefix.slice(prefix.indexOf('=') + 1, -1).toUpperCase();
```
[the remark this answers is not in the document]
Comment (BIMvoice) [reply packages/export/src/reference-collector.ts]: Confirmed as a real defect. Reproduced with a RED test (`relationship-filter-gate.test.ts`, a fixture whose line reads `#6 = IFCRELCONNECTSSTRUCTURALMEMBER(...)` — space after `=`) before the fix: the whole relationship was withheld and `#5=IFCAXIS2PLACEMENT3D` disappeared along with the member-to-connection association, instead of just rewriting `ConditionCoordinateSystem` to `$`. Fixed at the extraction site with `.trim()`, as proposed, in fe04ec31a3247a09aaeaaa26e71dbd926fb4deb0:

```diff
-  const entityType = prefix.slice(prefix.indexOf('=') + 1, -1).toUpperCase();
+  const entityType = prefix.slice(prefix.indexOf('=') + 1, -1).trim().toUpperCase();
```

Chose extraction-site `.trim()` over tightening the regex: the line regex's own capture (`\w+`) already excludes whitespace from the token itself — the only whitespace that can leak through is the prefix's own leading `\s*=\s*`, so trimming the sliced substring is a complete fix with no regex change needed.

Also caught and fixed the same whitespace gap in this test file's own `findDanglingRefs` helper (its defining-line regex was `#(\d+)=` with no `\s*`, which would have misreported the new test's own `#6 = ...` line as dangling).

Swept `reference-collector.ts` and neighbouring `export/src` files for the same shape (a value `slice`d out around a regex-permitted separator, then compared with `===`): this was the only `slice(...indexOf...)` extraction in the file, and `retype.ts`'s one `indexOf('=')` use doesn't extract a type token for comparison, so no other instance exists in this file's scope.

Also checked leading whitespace before `#`: the outer regex is anchored `^#\d+...` with no `\s*` before it, so a line with leading whitespace simply fails to match and is returned unchanged (the documented fallback) — not a comparison-defeating case.

GREEN after the fix: `packages/export` — 726 passed / 30 skipped (53 files), up from 725 passed before this fix (the new test only). Both `tsc --noEmit` (src) and `scripts/typecheck-tests.mjs` (tests) clean, oxlint clean on the changed files. Changeset added (`step-export-entity-type-token-whitespace.md`).
